### PR TITLE
feat(parser): allow parsing files with missing close braces

### PR DESCRIPTION
**Pull request by dsherret**
**Description:**

This is a small, but huge improvement for swc to get ASTs in spite of syntax errors. Essentially, it allows parsing even if there's a missing close brace in a section of the code.

For example, the following will now parse, but have two emitted diagnostics for missing `}`:

```ts
class MyClass {
  prop: string;

  myMethod() {
    console.log(5);
```
Part of what's outlined in #1178

**BREAKING CHANGE:**

A missing close brace token will no longer be a fatal parsing error, but will instead emit diagnostics.

### CODE REVIEW

File: crates/swc_ecma_parser/tests/span/js/decl/class-no-close-brace.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Test {
+
+  method() {}
```
Comment (dsherret) [inline crates/swc_ecma_parser/tests/span/js/decl/class-no-close-brace.js:3]: Waiting on #4186 in order to update the .js.spans below.

Comment (kdy1) [reply crates/swc_ecma_parser/tests/span/js/decl/class-no-close-brace.js:3]: Thank you!

Actually, I tried error recovery several time and failed, but I think I can do it if I try in some gradual way...
The problem was an infinite loop in some test cases

**Review by kdy1 — COMMENTED** [review]
Code changes LGTM.
Thank you!

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_parser

File: crates/swc_ecma_parser/tests/span/js/decl/class-no-close-brace.js (hunk 1 above)
Comment (dsherret) [reply crates/swc_ecma_parser/tests/span/js/decl/class-no-close-brace.js:3]: This is resolved now.

I think what would be good for error handling is to look at what TypeScript does for certain things and work incrementally on it. TypeScript seems to work well in the editor.

**Pull request merged**